types/mithril/index.d.ts

```diff
@@ -246,7 +246,7 @@ declare namespace Mithril {
 	 * Components are a mechanism to encapsulate parts of a view to make code easier to organize and/or reuse.
 	 * Any Javascript object that has a view method is a Mithril component. Components can be consumed via the m() utility.
 	 */
-	type Comp<Attrs, State extends Lifecycle<Attrs, State>> = Component<Attrs, State> & State;
```

Comment:
Apologies I haven't commented yet, finally had some time to look at this.

The changes are almost okay, but unfortunately we lose some type checking. For example, if a user attempts to add an `onremove` method to their `State` type, it should enforce the correct signature since it will be recognized by Mithril as a lifecycle hook:

```typescript
interface State {
    // Incorrect signature for this lifecycle hook (should take a Vnode parameter)
	onremove(i: number): number;
}

const c: Component<{}, State> = {
	oninit(vnode) {
		vnode.state.onremove = i => i + 1;
	},
	view(vnode) {
        return m('p', 'Test')
    }
};
```

The above will have compiler errors using our current types, but won't with the changes in this branch.

It would be good to add this to our tests somehow (but it would cause a failure.)